File: python/src/lru_cache.py
from collections import OrderedDict


# Using OrderedDict (C implementation) instead of manual implementation to optimise performance.
# There is a faster functools.lru_cache in Python that is a LRU cache implementation but no explicit get()/ put() access to the cache.
class LRUCache:
    def __init__(self, capacity):
        self.capacity = capacity
        self.cache = OrderedDict()

    def get(self, key: int):  # O(1)
        if key not in self.cache:
            return None

        # move the key to the end of the OrderedDict (MRU)
        self.cache.move_to_end(key)
        return self.cache[key]

    def put(self, key: int, value: int):  # O(1)
        if key in self.cache:
            # update the value
            self.cache[key] = value
            # move the key to the end of the OrderedDict (MRU)
            self.cache.move_to_end(key)
            return

        # if the cache is full, evict the least recently used key (LRU)
        if len(self.cache) >= self.capacity:
            # evict the least recently used key (LRU)
            self.cache.popitem(last=False)

        # insert the new key-value pair
        self.cache[key] = value

File: java/src/main/java/com/polyglot/LRUCache.java
package com.polyglot;

import java.util.LinkedHashMap;
import java.util.Map;

// Using LinkedHashMap instead of manual implementation to study real-world LRU cache implementation
// in Java.
public class LRUCache {
  private final int capacity;
  private final LinkedHashMap<Integer, Integer> map;

  public LRUCache(int capacity) {
    this.capacity = capacity;
    this.map =
        new LinkedHashMap<Integer, Integer>(capacity, 0.75f, true) {
          // Override the removeEldestEntry method to remove the least recently used entry when the
          // cache is full
          @Override
          protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
            return size() > LRUCache.this.capacity;
          }
        };
  }

  public Integer get(Integer key) {
    return map.getOrDefault(key, null);
  }

  public void put(Integer key, Integer value) {
    map.put(key, value);
  }

  public int size() {
    return map.size();
  }

  public int capacity() {
    return capacity;
  }
}
